Feature/remove player usecase

AngrySaltyFish24 opened the pull request:
Resolves https://github.com/stefanoberri/roachcase/issues/19

This pr adds the ability for the user to remove a player by name from the roachcase.

At present no errors are raised if a user with that name does not exist. This can be dealt with in another PR.

Tests have been added and are passing, but unfortunately the DynamoDB test is running on make test despite being marked as a livedb test. Perhaps you can spot what i am doing wrong here.

AngrySaltyFish24 committed:
Removing print statement

AngrySaltyFish24 committed:
Expanding the tutorial documentation to include `remove_player`

stefanoberri commented:
Hi @AngrySaltyFish24. Thank you for contributing! To keep the codebase clean, I have rebased my branch in one commit and merged into develop. Going forward, I will pay more attention to branches, and we should always branch off develop.
Please change the target of your PR onto develop now (it might require rebasing). When I merge, I will Squash your contribution into a single commit, and as text for the commit I will use the first comment of the PR, so please update it as you go, following these guidelines:

https://medium.com/compass-true-north/writing-good-commit-messages-fc33af9d6321

If there is something not clear, please let me know.

AngrySaltyFish24 committed:
This adds the remove player usecase.

This commit adds code to allow a user to remove a player from the
roachcase using the api. At current no error handling has been added.

stefanoberri commented on roachcase/_infrastructure/dynamodb_repositories.py:
```diff
@@ -20,6 +20,30 @@ class DynamoDBGateway:
         self.__client = boto3.client("dynamodb")
 
+<<<<<<< HEAD
+=======
```
There are some conflicts to solve

AngrySaltyFish24 replied:
Yh sorry my bad, did a rebase to remove

stefanoberri commented on roachcase/_infrastructure/dynamodb_repositories.py:
```diff
@@ -41,6 +65,13 @@ def create_table(self, specs: Dict[str, Any]) -> None:
             waiter = self.__client.get_waiter("table_exists")
             waiter.wait(TableName=table_name)
 
+    def remove(self, table_name: str, item: Dict[str, Any]) -> None:
+        """Add an item to a table"""
+        try:
+            self.__client.delete_item(TableName=table_name, Key=item)
+        except botocore.exceptions.ClientError as error:
+            self.__handle_error(error)
```
This requires handling the tests correctly. There are two reasons - at least - why it raises an error: 
- The table does not exist
- The item does not exist
The in memory and the DynamoDB gateway should behave the same
Maybe this requires a bit more work on my side, or we need to find a way to run these tests at integration.

AngrySaltyFish24 replied:
as per the comment down below, error handling will be done in a future PR.

stefanoberri commented on tests/test_api.py:
```diff
@@ -14,14 +14,31 @@ def test_set_persistence():
 
-def test_add_list_players():
+@pytest.fixture
+def mem_persistance():
```
the `set_persistence("memory") reset the repositories, like starting from scratch, we need to do that every time, or different tests will know about previous state, I think. This is a bit tricky, and maybe this is not how we want it to work...

AngrySaltyFish24 replied:
I think this fixture should be rename to something like `clear_memory` and have it call `set_persistence("memory")` once. The we set the fixture to autouse (https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request) to make sure each test is starting from a clean slate. I think this would be a good solution

stefanoberri replied:
I like the idea of the autouse to make sure we start from a clean persistence state.

AngrySaltyFish24 replied:
I quickly implemented it in d6e2951, let me know what you think

stefanoberri commented on tests/test_api.py:
```diff
@@ -14,14 +14,31 @@ def test_set_persistence():
+        assert set(observed) == set(["Alice", "Bob"])
+
+    def check_remove_players(self):
+        roachcase.remove_player("Alice")
```
This test, for instance, is strange. How can we remove "Alice" if we did not add it? I think this works because the persistence "survives" across tests, but that's not what we want - I think.

AngrySaltyFish24 replied:
This was done by intention. In order for us to remove a player, a player must exist in the roachcase and to check the contents we need to use the list method. This means the testing of a feature depends upon the functionality of another feature. This means that we need to test these features one after the other in the same test, i figured a good way to structure this was to have checks that the test called, largely inspired by https://github.com/pytest-dev/pytest/issues/2317#issuecomment-317438952. An alternative could be to use the `pytest-dependency` plugin

stefanoberri commented on tests/test_api.py:
```diff
@@ -14,14 +14,31 @@ def test_set_persistence():
+        observed = roachcase.list_players()
+        assert set(observed) == set(["Bob"])
+
+    @pytest.mark.usefixtures("mem_persistance")
```
Not a big fan of `pytest.mark.usefixtures`. We should use the fixture itself. 
And not sure what this test does. "test management"? What management?

AngrySaltyFish24 replied:
I've removed the use of mark in favour of the fixture itselt and renamed the test in d6e2951. Let me know your thoughts

stefanoberri commented on tests/test_repositories.py:
```diff
@@ -16,10 +16,94 @@ def test_round_trip(self, john, repo):
 
+        repo.remove(john)
+        observed = repo.get()
+        assert list(observed) == []
```
Here we could try and remove `john` again. What should happen?

AngrySaltyFish24 replied:
It should error. I plan to implement this but thought it would be better to get feedback early one to see if im heading right direction with the implementation. I can of course add it to this PR if you like or i could create a seperate issue and PR. The choice is yours!

stefanoberri replied:
OK. Happy to wait.

stefanoberri commented on roachcase/_repositories.py:
```diff
@@ -8,11 +8,99 @@ class PlayerAlreadyExistError(ValueError):
+
+    @abc.abstractmethod
+    def remove(self, table_name: str, item: Dict[str, Any]) -> None:
+        """Remove an item from a `table_name`"""
```
I think this is duplication of the `delete` method.

stefanoberri reviewed (changes requested):
I have messed up with the branches and rebases. Sorry. I will try and fix that. Meanwhile have a look at some of my comments.

AngrySaltyFish24 committed:
Adding a test for removing an item

The test simply adds an item, removes it and checks that the list is
empty.

Will help make the remove usecase more maintable code.

stefanoberri commented on roachcase/_infrastructure/shelf_repositories.py:
```diff
@@ -31,6 +31,13 @@ def get(self) -> Iterable[_entities.Player]:
         return result
 
+    def remove(self, player: _entities.Player) -> None:
+        """Get players from the repository"""
```
Update docstring.

stefanoberri commented on roachcase/_infrastructure/dynamodb_repositories.py:
```diff
@@ -41,6 +41,13 @@ def create_table(self, specs: Dict[str, Any]) -> None:
             waiter = self.__client.get_waiter("table_exists")
             waiter.wait(TableName=table_name)
 
+    def remove(self, table_name: str, item: Dict[str, Any]) -> None:
```
This is duplicated of `delete`. I don't mind which one we use, but we should have only one of them.

stefanoberri reviewed (commented):
Some more comments, I understand some of your previous points. Thanks.

AngrySaltyFish24 committed:
Cleans up naming and fixtures in `test_api`

The following changes:
- Renamed `test_management` -> `test_player_management` as the previous
  name was ambigious
- Refactored `mem_persistance` fixture to `clear_memory`. This fixture
  is auto used for every test and sets the memory persistance once so
  each test starts with a blank state making them more reproducable.
